tf_quant_finance/volatility/barrier_option.py

```diff
@@ -0,0 +1,367 @@
+        [0., 0., 1., 1., 0., 0., -1., -1., 0., 0., 1., 1.])
+
+    # Constructing Masks
+    @tf.function
+    def get_out_map(params):
+      """
+      Function maps params to option pricing masks
+      Args:
+      params: Tuple of tensors. (barrier price, strike_price, option type)
+        params[0] = barrier_price
+        params[1] = strike_price
+        params[2] = otype
+      Returns:
+      returns mask used to price specified option
+      """
+      out_map = down_and_in_call_tensor_greater_strike # Default
+      if tf.math.less(params[1], params[0]):
+        if tf.equal(params[2], down_and_in_call):
+          out_map = down_and_in_call_tensor_lower_strike
+        if tf.equal(params[2], down_and_in_put):
+          out_map = down_and_in_put_tensor_lower_strike
+        if tf.equal(params[2], up_and_in_call):
+          out_map = up_and_in_call_tensor_lower_strike
+        if tf.equal(params[2], up_and_in_put):
+          out_map = up_and_in_put_tensor_lower_strike
+        if tf.equal(params[2], down_and_out_call):
+          out_map = down_and_out_call_tensor_lower_strike
+        if tf.equal(params[2], down_and_out_put):
+          out_map = down_and_out_put_tensor_lower_strike
+        if tf.equal(params[2], up_and_out_call):
+          out_map = up_and_out_call_tensor_lower_strike
+        if tf.equal(params[2], up_and_out_put):
+          out_map = up_and_out_put_tensor_lower_strike
+      else:
+        if tf.equal(params[2], down_and_in_call):
+          out_map = down_and_in_call_tensor_greater_strike
+        if tf.equal(params[2], down_and_in_put):
+          out_map = down_and_in_put_tensor_greater_strike
+        if tf.equal(params[2], up_and_in_call):
+          out_map = up_and_in_call_tensor_greater_strike
+        if tf.equal(params[2], up_and_in_put):
+          out_map = up_and_in_put_tensor_greater_strike
+        if tf.equal(params[2], down_and_out_call):
+          out_map = down_and_out_call_tensor_greater_strike
+        if tf.equal(params[2], down_and_out_put):
+          out_map = down_and_out_put_tensor_greater_strike
+        if tf.equal(params[2], up_and_out_call):
+          out_map = up_and_out_call_tensor_greater_strike
+        if tf.equal(params[2], up_and_out_put):
+          out_map = up_and_out_put_tensor_greater_strike
+      return out_map
+    
+    @tf.function
+    def get_below_or_above(x):
+      if tf.math.less(x[0], x[1]):
+        return -1.
+      return 1.
+
+    @tf.function
+    def get_call_or_put(x):
+      if tf.equal(tf.math.mod(x, 2), 0):
+        return -1.
+      return 1.
+
+    # build call or put, below or above barrier price, and masks
+    if otype.shape == ():
+      below_or_above = get_below_or_above((asset_price, barrier_price))
+      call_or_put = get_call_or_put(otype)
+      masks = get_out_map((barrier_price, strike_price, otype))
+    else:
+      masks = tf.map_fn(
+          get_out_map,
+          (barrier_price, strike_price, otype), dtype=tf.float32)
```

Comment:
no need for map_fn. tf.where should do the job and it will be much faster

Reply:
Done!